be/src/olap/memtable_flush_executor.h

```diff
@@ -0,0 +1,158 @@
+class ExecEnv;
+class MemTable;
+
+// The context for a memtable to be flushed.
+class FlushHandler;
+struct MemTableFlushContext {
+    // generated by memtable flush executor
+    // denote a certain memtable flush execution.
+    int64_t flush_id;
+    // memtable to be flushed
+    std::shared_ptr<MemTable> memtable;
+    // flush handler from a delta writer.
+    // use shared ptr because flush_handler may be deleted before this
+    // memtable being flushed. so we need to make sure the flush_handler
+    // is alive until this memtable being flushed.
+    std::shared_ptr<FlushHandler> flush_handler;
+};
+
+// the flush result of a single memtable flush
+struct FlushResult {
+    OLAPStatus flush_status;
+    int64_t flush_time_ns;
+};
+
+// the statistic of a certain flush handler.
+// use atomic because it may be updated by multi threads
+struct FlushStatistic {
+    std::atomic_int64_t flush_time_ns = {0};
+    std::atomic_int64_t flush_count= {0};
+};
+std::ostream& operator<<(std::ostream& os, const FlushStatistic& stat);
+
+// flush handler is for flushing memtables in a delta writer
+// This class must be wrapped by std::shared_ptr, or you will get bad_weak_ptr exception
+// when calling submit();
+class MemTableFlushExecutor;
+class FlushHandler : public std::enable_shared_from_this<FlushHandler> {
+public:
+    FlushHandler(int32_t flush_queue_idx, MemTableFlushExecutor* flush_executor):
+        _flush_queue_idx(flush_queue_idx),
+        _last_flush_status(OLAP_SUCCESS),
+        _flush_executor(flush_executor) {
+    }
+
+    // submit a memtable to flush. return error if some previous submitted MemTable has failed  
+    OLAPStatus submit(std::shared_ptr<MemTable> memtable);
+    // wait for all submitted memtable finished.
+    OLAPStatus wait();
+    // get flush operations' statistics
+    const FlushStatistic& get_stats();
+    // called when a memtable is finished by executor.
+    void on_flush_finished(const FlushResult& res);
+
+    OLAPStatus last_flush_status() { return _last_flush_status.load(); }
+private:
+    // flush queue idx in memtable flush executor
+    int32_t _flush_queue_idx;
+    // the flush status of last memtable
+    std::atomic<OLAPStatus> _last_flush_status;
+    // save all futures of memtable flush execution.
+    std::queue<std::future<OLAPStatus>> _flush_futures;
+    FlushStatistic _stats;
+    MemTableFlushExecutor* _flush_executor;
+};
+
+// MemTableFlushExecutor is for flushing memtables to disk.
+// Each data directory has a specified number of worker threads and a corresponding number of flush queues.
+// Each worker thread only takes memtable from the corresponding flush queue and writes it to disk.
+// User SHOULD NOT call method of this class directly, use pattern should be:
+//
+//      ...
+//      FlushHandler* flush_handler;
+//      memTableFlushExecutor.create_flush_handler(path_hash, &flush_handler);
+//      std::shared_ptr<FlushHandler> shared_handler(flush_handler);
```

Comment:
If client should wrap returned pointer with a shared_ptr, why not define the argument with `std::shared_ptr<>*`

Reply:
OK